tests/integration/api/users/GroupSearchTest.php

```diff
@@ -0,0 +1,331 @@
+<?php
+
+/*
+ * This file is part of Flarum.
+ *
+ * For detailed copyright and license information, please view the
+ * LICENSE file that was distributed with this source code.
+ */
+
+namespace Flarum\Tests\integration\api\users;
+
+use Flarum\Tests\integration\RetrievesAuthorizedUsers;
+use Flarum\Tests\integration\TestCase;
+
+class GroupSearchTest extends TestCase
+{
+    use RetrievesAuthorizedUsers;
+
+    public function setUp(): void
+    {
+        parent::setUp(); // TODO: Change the autogenerated stub
+
+        $this->prepareDatabase([
+            'users' => [
+                $this->normalUser(),
+            ],
+        ]);
+    }
+
+    /**
+     * @test
+     */
+    public function disallows_group_filter_for_user_without_permission()
+    {
+        $response = $this->send(
+            $this->request('GET', '/api/users')
+                ->withQueryParams(['filter' => ['q' => 'group:admin']])
+        );
+
+        $this->assertEquals(403, $response->getStatusCode());
+    }
+
+    /**
+     * @test
+     */
+    public function allows_group_filter_for_admin()
+    {
+        $response = $this->send(
+            $this->request('GET', '/api/users?filter%5Bq%5D=group%3Aadmin', [
+                'authenticatedAs' => 1,
+            ])->withQueryParams(['filter' => ['q' => 'group:admin']])
+        );
+
+        $this->assertEquals(200, $response->getStatusCode());
+    }
+
+    /**
+     * @test
+     */
+    public function allows_group_filter_for_user_with_permission()
+    {
+        $this->prepareDatabase([
+            'group_permission' => [
+                ['permission' => 'viewUserList', 'group_id' => 2],
+            ],
+        ]);
+        $response = $this->send(
+            $this->request('GET', '/api/users')
+                ->withQueryParams(['filter' => ['q' => 'group:admin']])
+        );
+
+        $this->assertEquals(200, $response->getStatusCode());
+    }
+
+    /**
+     * @test
+     */
+    public function non_admin_gets_correct_results_group_name_singular()
+    {
+        $this->prepareDatabase([
+            'group_permission' => [
+                ['permission' => 'viewUserList', 'group_id' => 2],
+            ],
+        ]);
+
+        $response = $this->send(
+            $this->request('GET', '/api/users')
+                ->withQueryParams(['filter' => ['q' => 'group:admin']])
+        );
+        $responseBodyContents = json_decode($response->getBody()->getContents());
```

Comment:
Why are we decoding this, then immediately re-encoding it?

Reply:
It's decoded so that we can access the object properties, and then encoded for the message part of the assert so that if the test fails, you can see what was returned in the output.
I am happy to remove it if it isn't required though.

Reply:
Ah I see, I like that! We should do that more actually...